### ISSUE TITLE
Docker not prompting user for auth on pull / making bad decision to fallback to /v1

### ISSUE THREAD

**Issue by JonathonReinhart** [docker/docker#18708]
**Summary**
"Error: image not found" on `docker pull` when using private registry and token auth. It seems to be a login / authorization problem, because `docker login` "fixes" it.

This is a *huge* problem for me, because my environment forbids me from saving LDAP credentials in an unencrypted file on disk (the notorious #10318). I was planning on using the work-around I described [here](https://github.com/docker/docker/issues/10318#issuecomment-152017414), which is `ln -sf /dev/null ~/.docker/config.json`. However, this would prevent me from ever being able to `pull` an image from my private registry.

**Related**
This seems similar to #15759, but there's a not of chaff there, [regarding S3](https://github.com/docker/docker/issues/15759#issuecomment-147355492), and [tdterry's comment](https://github.com/docker/docker/issues/15759#issuecomment-155150663) shows logs that don't match mine.

Also possibly related to docker/distribution#884.

**Configuration / Versions**
Docker: `1.8.2-fc22`
Private registry (local file storage): `registry github.com/docker/distribution v2.1.1`
[Portus](http://port.us.org/) (token auth server): `Portus-20151027090209-1.1.x86_64`

**Problem Demonstration**

Clear auth:
```
$ rm ~/.docker/config.json
```

Try to pull:
```
$ docker pull docker-reg.example.com/jreinhart/busybox:1.0
Trying to pull docker-reg.example.com/jreinhart/busybox ... not found
Error: image jreinhart/busybox:1.0 not found
```

Registry Log (abbreviated):
```
GET /v2/  --> 401 "authorization token required"
GET /v2/  --> 401 "authorization token required"
GET /v1/repositories/jreinhart/busybox/images --> 404
```

Portus log (abbreviated):
```
Started GET "/v2/token?scope=repository%3Ajreinhart%2Fbusybox%3Apush%2Cpull&service=docker-reg.example.com" for 192.168.1.158 at 2015-12-16 10:17:13 -0500
Processing by Api::V2::TokensController#show as JSON
Parameters: {"scope"=>"repository:jreinhart/busybox:push,pull", "service"=>"docker-reg.example.com"}
Denied access on these grounds: not allowed to  this nil
Completed 401 Unauthorized in 10ms (ActiveRecord: 1.3ms)
```

**Description of Problem(s)**
- Docker should have aborted when it got the 401 from Portus (the auth server).
- Docker should have prompted the user for creds when it realized `~/.docker/config.json` didn't have creds for the given auth server.
- Docker should not have fallen back to `/v1` in this case (because `/v2` did not 404).
- Docker should not have given an incorrect error message like `Error: image jreinhart/busybox:1.0 not found`.

Simply running `docker login` makes `docker pull` work again, but this of course requires saving my network creds unencrypted, which is unacceptable.

It's kind of mind-boggling that such an obvious scenario behaves so poorly.

**cc** @stevvooe @dmp42

**Comment by JonathonReinhart** [comment]
@GordonTheTurtle Yes, I know you're a bot. Yes I know there are thousands of open issues against Docker, and this is certainly an effort to mitigate that.  My bug report was extremely thorough, so other than adding the text **BUG REPORT INFORMATION**, I'm ignoring this.

**Comment by stevvooe** [comment]
@JonathonReinhart Calm down. No sense yelling at the bot. We just delete the Gordon messages if they don't make sense.

The fallback situation is non-ideal. We are slowly phasing out v1 support, so the fallbacks are becoming less and less aggressive as this happens.

In this particular case, one cannot tell the difference between a 401 not being authorized and 401 not authorized because v2 is not supported. Yes, arguably `/v2/` returning a 401 should be considered a `/v2/` endpoint, especially if there is a version header present. This is problematic when running a dual v1/v2 setup, if I remember correctly.

@aaronlehmann Any thoughts here?

**Comment by aaronlehmann** [comment]
If I understand correctly, this isn't about overly aggressive v1 fallback as much as a feature request for `docker pull` and `docker push` to prompt for credentials if none were provided previously with `docker login`. I don't think that would be very hard to implement, assuming there's no controversy about turning `docker pull` and `docker push` into possibly interactive commands.

**Comment by JonathonReinhart** [comment]
@aaronlehmann `docker push` is already interactive:

```
$ docker push docker-reg.example.com/jreinhart/busybox
The push refers to a repository [docker-reg.example.com/jreinhart/busybox] (len: 1)
fc0db02f3072: Image push failed 

Please login prior to push:
Username: jreinhart
Password: 
Email: jreinhart@example.com
WARNING: login credentials saved in /home/jreinhart/.docker/config.json
Login Succeeded
The push refers to a repository [docker-reg.example.com/jreinhart/busybox] (len: 1)
fc0db02f3072: Image already exists 
5c5fb281b01e: Image already exists 
8.0: digest: sha256:ddd04870d28de86975649ca115f037c31ad9816506e1317af790bd325e306357 size: 3213
```

**Comment by aaronlehmann** [comment]
I didn't realize that. In that case, `pull` should definitely behave similarly.

**Comment by ekristen** [comment]
I've got about 40 servers pulling from a v2 registry, and one refuses to. Basically it starts to pull all the images and at some point it checks /v2/ gets the 401 and then checks /v1/ and gets 404 and then stops downloading all image layers and states that it is not found. 

I cannot reproduce on any other system at the moment (thankfully)

`Docker version 1.9.1, build a34a1d5`

**Comment by aaronlehmann** [comment]
@ekristen: I'd suggest filing a separate issue for this. If you are able to reproduce it, running the Docker daemon with `-D` (debug) and capturing the daemon log during one of these pulls would be helpful.

**Comment by LK4D4** [comment]
@aaronlehmann Is this still an issue?

**Comment by aaronlehmann** [comment]
I think so. But the issue author says:

> This is a huge problem for me, because my environment forbids me from saving LDAP credentials in an unencrypted file on disk (the notorious #10318)

and now there are workaround for this (identity tokens, and registry credential stores)

**Comment by LK4D4** [comment]
@aaronlehmann I still think that we should have that for symmetry. Leaving open.

**Comment by thaJeztah** [comment]
@aaronlehmann would https://github.com/docker/docker/pull/26354 help for this?

**Comment by aaronlehmann** [comment]
No, I don't think it's really relevant. As I understand that PR, it's a tweak to the credential store functionality that avoids the need for Docker to keep track of which registries the user has logged into. The existing credential store functionality already makes it possible to avoid storing passwords in plaintext.

**Comment by KingEmet** [comment]
@thaJeztah @aaronlehmann is correct, my change only alters the behavior of native stores.
